esmvalcore/preprocessor/_bias.py

```diff
@@ -40,79 +66,123 @@ def bias(products, bias_type='absolute', denominator_mask_threshold=1e-3,
+        return CubeList(cubes)
+
+    # Otherwise, iterate over all input products, calculate bias and adapt
+    # metadata and provenance information accordingly
     output_products = set()
     for product in products:
-        if product == reference_product:
+        if product == ref_product:
             continue
         cube = concatenate(product.cubes)
-        cube_metadata = cube.metadata
 
         # Calculate bias
-        if bias_type == 'absolute':
-            cube = cube - ref_cube
-            new_units = str(cube.units)
-        elif bias_type == 'relative':
-            cube = (cube - ref_cube) / ref_cube
-            new_units = '1'
-        else:
-            raise ValueError(
-                f"Expected one of ['absolute', 'relative'] for bias_type, got "
-                f"'{bias_type}'")
+        cube = _calculate_bias(cube, ref_cube, bias_type)
 
-        # Adapt cube metadata and provenance information
-        cube.metadata = cube_metadata
-        cube.units = new_units
-        product.attributes['units'] = new_units
-        product.wasderivedfrom(reference_product)
+        # Adapt metadata and provenance information
+        product.attributes['units'] = str(cube.units)
+        if ref_product is not None:
+            product.wasderivedfrom(ref_product)
 
-        product.cubes = iris.cube.CubeList([cube])
+        product.cubes = CubeList([cube])
         output_products.add(product)
 
     # Add reference dataset to output if desired
-    if keep_reference_dataset:
-        output_products.add(reference_product)
+    if keep_reference_dataset and ref_product is not None:
+        output_products.add(ref_product)
 
     return output_products
+
+
+def _get_ref(products, ref_tag: str) -> tuple[Cube, PreprocessorFile]:
+    """Get reference cube and product."""
+    ref_products = []
+    for product in products:
+        if product.attributes.get(ref_tag, False):
+            ref_products.append(product)
+    if len(ref_products) != 1:
+        raise ValueError(
+            f"Expected exactly 1 dataset with '{ref_tag}: true', found "
+            f"{len(ref_products):d}"
+        )
+    ref_product = ref_products[0]
+
+    # Extract reference cube
+    # Note: For technical reasons, product objects contain the member
+    # ``cubes``, which is a list of cubes. However, this is expected to be a
+    # list with exactly one element due to the call of concatenate earlier in
+    # the preprocessing chain of ESMValTool. To make sure that this
+    # preprocessor can also be used outside the ESMValTool preprocessing chain,
+    # an additional concatenate call is added here.
+    ref_cube = concatenate(ref_product.cubes)
+
+    return (ref_cube, ref_product)
+
+
+def _calculate_bias(cube: Cube, ref_cube: Cube, bias_type: BiasType) -> Cube:
+    """Calculate bias for a single cube relative to a reference cube."""
+    cube_metadata = cube.metadata
+
+    if bias_type == 'absolute':
+        cube = cube - ref_cube
```

Comment:
does iris now allow cube subtractions when cubes have differing metadata?

Reply:
no, that what you mentioned. There was a time in the past when iris was disallowing cube subtractions even if it rained outside